### PR TITLE
SHMEM_LOCKS: MCS implementation of SHMEM LOCKS

### PR DESCRIPTION
**Pull request by vvenkates27**
Adding MCS algorithm-based implementation for shmem_locks to improve performance for large scale SHMEM applications using locks. MCS lock is now the default algorithm, use the following MCA parameter to disable.
-- Use MCA parameter --mca oshmem_enable_mcs_lock 0 to disable mcs locks and revert to default ticket locking.
-- Use oshmem_api_verbose 10 for debug information on shmem_locks. 
@manjugv @brminich @rakhmets @yosefe

### CODE REVIEW

File: oshmem/shmem/c/shmem_test_lock.c [hunk 1]
```diff
@@ -28,5 +29,11 @@
 
 int shmem_test_lock(volatile long *lock)
 {
-    return _shmem_test_lock((void *)lock, sizeof(long));
+	if (oshmem_shmem_mcs_locks) {
+		if (oshmem_shmem_lock_debug)
+			printf("Calling MCS implementation\n");
```
Comment (bosilca) [inline oshmem/shmem/c/shmem_test_lock.c:34]: Please use OPAL_OUTPUT_VERBOSE or other OMPI debugging support, but not printf.

Comment (vvenkates27) [reply oshmem/shmem/c/shmem_test_lock.c]: Thanks for pointing out. I have updated and pushed a new patch.

File: oshmem/shmem/c/shmem_clear_lock.c [hunk 2]
```diff
@@ -27,5 +28,10 @@
 
 void shmem_clear_lock(volatile long *lock)
 {
-    _shmem_clear_lock((void *)lock, sizeof(long));
+	if (oshmem_shmem_mcs_locks) {
```
Comment (brminich) [inline oshmem/shmem/c/shmem_clear_lock.c:31]: general comment: pls use 4 space tabs as per [OMPI Coding Style](https://docs.open-mpi.org/en/main/developers/source-code.html#code-style)

File: oshmem/shmem/c/shmem_clear_lock.c [hunk 3]
```diff
@@ -27,5 +28,10 @@
...
+		SHMEM_API_VERBOSE(10, "shmem_clear_lock using MCS Lock");
```
Comment (brminich) [inline oshmem/shmem/c/shmem_clear_lock.c:32]: maybe move this print to the very beggining of the func and also print locking scheme being cleaned?

File: oshmem/shmem/c/shmem_fetch.c [hunk 4]
```diff
@@ -45,7 +45,8 @@
     } while (0)
 
 #define SHMEM_CTX_TYPE_ATOMIC_FETCH(type_name, type, prefix)        \
-    type prefix##_ctx##type_name##_atomic_fetch(shmem_ctx_t ctx, const type *target, int pe) \
+    type prefix##_ctx##type_name##_atomic_fetch(shmem_ctx_t ctx,	\
+						const type *target, int pe) \
```
Comment (brminich) [inline oshmem/shmem/c/shmem_fetch.c:49]: unrelated change, pls remove

File: oshmem/shmem/c/shmem_lock.c [hunk 5]
```diff
@@ -269,20 +269,24 @@ static uint64_t shmem_lock_cswap(void *target,
     uint64_t prev_value = 0;
 
     if (target_size == 8) {
-        MCA_ATOMIC_CALL(cswap(oshmem_ctx_default, target, (void*)&prev_value, cond, value, target_size, pe));
+        MCA_ATOMIC_CALL(cswap(oshmem_ctx_default, target, (void*)&prev_value,
+			      cond, value, target_size, pe));
```
Comment (brminich) [inline oshmem/shmem/c/shmem_lock.c:273]: pls consider pushing changes in this file in a separate PR, as they are not relavant for the new locking algo

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 6]
```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright (c) 2013      Mellanox Technologies, Inc.
+ *                         All rights reserved.
+ * Copyright (c) 2014      Cisco Systems, Inc.  All rights reserved.
+ * Copyright (c) 2015      Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
+ * $COPYRIGHT$
+ *
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:8]: as this is newly introduced file, i'd guess only relevant Nvidia copyright notice should be present

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+	RUNTIME_CHECK_RC(rc);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+	/** Signal for setting lock */
+	new_tail_req = SET_SIGNAL(my_pe);
+	/**
+	 * Swap and make me the new tail and update in tail owner
+	 * Get the previous tail PE.
+	 * int prev_tail	 = shmem_atomic_swap(tail, new_tail_req,
+				     mcs_tail_owner);
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:106]: what does this comment say?

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+			  mcs_tail_owner));
+	RUNTIME_CHECK_RC(rc);
+	prev_tailpe	= GET_PE(prev_tail);
+
+	if (SIGNAL(prev_tail)) {
+		/** Someone else has got the lock before this PE
+		 *  Adding this PE to the previous tail PE's Next pointer
+		 *  Substract the SIGNAL Bit to avoid changing it.
+		 *  shmem_atomic_add(next, my_pe - NEXT_MASK,
+				     prev_tailpe);
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:123]: why need this?

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+		 **/
+		rc = MCA_ATOMIC_CALL(add(oshmem_ctx_default,
+					 (void *)next, SIGNAL_MASK,
+					 sizeof(int), my_pe));
+		RUNTIME_CHECK_RC(rc);
+		MCA_SPML_CALL(quiet(oshmem_ctx_default));
+		/**
+		 * Wait for predecessor release lock to this PE
+		 * signal to false.
+		 * int curr = shmem_atomic_fetch(next, my_pe);
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:145]: why needed?

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+				  tval, sizeof(int), my_pe));
+        RUNTIME_CHECK_RC(rc);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+	if (next_value == NEXT_MASK) {
+		swap_cond = SET_SIGNAL(my_pe);
+
+		/**
+		 * prev_value = shmem_atomic_compare_swap(tail, swap_cond, 0,
+		 *					   mcs_tail_owner);
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:200]: isn't it almost the same what is used below?
same comment is relevant for other similar comments in this file

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+void _shmem_mcs_set_lock(long *lock) {
+	shmem_mcs_internal_set_lock(lock);
+}
+
+void _shmem_mcs_clear_lock(long *lock) {
+	shmem_mcs_internal_clear_lock(lock);
+}
+
+int  _shmem_mcs_test_lock(long *lock) {
+	return shmem_mcs_internal_test_lock(lock);
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:299]: why do we need these wrappers?

File: oshmem/shmem/c/shmem_set_lock.c [hunk 12]
```diff
@@ -27,5 +28,10 @@
 
 void shmem_set_lock(volatile long *lock)
 {
-    _shmem_set_lock((void *)lock, sizeof(long));
+	if (oshmem_shmem_mcs_locks) {
+		SHMEM_API_VERBOSE(10, "shmem_set_lock using MCS Lock");
```
Comment (brminich) [inline oshmem/shmem/c/shmem_set_lock.c:32]: maybe such comments are worth to be common for all type of locks? (i.e. to move them to the very beginning of func)

**Review by brminich — COMMENTED** [review]
👍 
have some minor comments

File: oshmem/shmem/c/shmem_clear_lock.c (hunk 3 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_clear_lock.c:32]: I have moved the print and also printing the locking scheme now with the latest commit

File: oshmem/shmem/c/shmem_fetch.c (hunk 4 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_fetch.c]: Done

File: oshmem/shmem/c/shmem_lock.c (hunk 5 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_lock.c]: Removed unwanted changes from this PR.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 6 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Fixed the License

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 7 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Artifact of previous implementation removed.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 10 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: I have removed this comment.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 11 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Was intended to keep similar to the existing implementation. Doesn't have purpose for this implementation. Removing it.

File: oshmem/shmem/c/shmem_set_lock.c (hunk 12 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_set_lock.c]: Agreed and moved in the latest commit.

**Comment by vvenkates27** [comment]
> 👍 have some minor comments

Thank you. Fixed all the issues pointed in the comment and uploaded a new patch. Please review and let me know if its ready for merging.

File: oshmem/shmem/c/shmem_clear_lock.c (hunk 2 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_clear_lock.c:34]: I have checked and fixed all files updated in this PR to stick to 4 space tabs.

Comment (brminich) [reply oshmem/shmem/c/shmem_clear_lock.c:34]: can you please double check? E.g. looks like 8-width tabs are used in this file (and possibly others)

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 13]
```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (c) 2023      NVIDIA Corporation.
+ *                         All rights reserved.
+ *                         and Technology (RIST). All rights reserved.
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:4]: leftover?

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 14]
```diff
@@ -0,0 +1,256 @@
...
+		  * Adding this PE to the previous tail PE's Next pointer
+		  * Substract the SIGNAL Bit to avoid changing it.
+		  **/
+		tmp_val = my_pe - NEXT_MASK;
+		retv = MCA_ATOMIC_CALL(add(oshmem_ctx_default, (void*)next, tmp_val,
+							 sizeof(int), prev_tailpe));
+		RUNTIME_CHECK_RC(retv);
+
+		/**
+		 *	This value to be changed eventually by predecessor
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:121]: alignment

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+_shmem_mcs_clear_lock(long *lockp)
+{
+	lock_t	*lock = (lock_t *) lockp;
+	int	mcs_tail_owner = MCSQ_TAIL_OWNER(lock);
+	int	*tail = &(lock->tail);
+	int	*next = &(lock->next);
+	int	my_pe = shmem_my_pe();
+	int	next_value	= 0;
+	int	swap_cond	= 0;
+	int	prev_value	= 0;
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:160]: no need for extra spaces before `=`

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+				 (void *)next, a_val,
+				 sizeof(a_val), NEXT(next_value)));
+	RUNTIME_CHECK_RC(retv);
+	MCA_SPML_CALL(quiet(oshmem_ctx_default));
+}
+
+int
+_shmem_mcs_test_lock(long *lockp)
+{
+	lock_t *lock = (lock_t *) lockp;
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:224]: pls align by `=` for consistency with other vars in this func

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+				   (uint64_t *)&(prev_tail),
+				   OSHMEM_ATOMIC_PTR_2_INT(&tmp_cond,
+					sizeof(tmp_cond)),
+				   OSHMEM_ATOMIC_PTR_2_INT(&new_tail_req,
+					sizeof(new_tail_req)),
+				   sizeof(int), mcs_tail_owner));
+	RUNTIME_CHECK_RC(retv);
+
+	if (0 == prev_tail)
+		return 0; /** lock acquired successfully */
```
Comment (brminich) [inline oshmem/shmem/c/shmem_mcs_lock.c:254]: can you please use {} for every statement, like
```suggestion
	if (0 == prev_tail) {
		return 0; /** lock acquired successfully */
         }
```
(we try to use this style for oshmem and ucx pml code)
also relevant for other one line statements in this PR

File: oshmem/shmem/c/shmem_set_lock.c [hunk 18]
```diff
@@ -27,5 +29,12 @@
 
 void shmem_set_lock(volatile long *lock)
 {
-    _shmem_set_lock((void *)lock, sizeof(long));
+	SHMEM_API_VERBOSE(10, "shmem_set_lock");
```
Comment (brminich) [inline oshmem/shmem/c/shmem_set_lock.c:32]: maybe remove this one now?

File: oshmem/shmem/c/shmem_test_lock.c [hunk 19]
```diff
@@ -28,5 +31,12 @@
 
 int shmem_test_lock(volatile long *lock)
 {
-    return _shmem_test_lock((void *)lock, sizeof(long));
+	SHMEM_API_VERBOSE(10, "shmem_test_lock");
```
Comment (brminich) [inline oshmem/shmem/c/shmem_test_lock.c:34]: remove?

Comment (vvenkates27) [reply oshmem/shmem/c/shmem_test_lock.c:34]: Done

File: oshmem/shmem/c/shmem_set_lock.c (hunk 18 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_set_lock.c]: Done.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 17 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Fixed.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 16 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Done.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 15 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Removed.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 14 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:120]: I have redone the alignment for the whole file now, hopefully this patch fixes the alignment issues.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 13 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Yes it was, removed now.

**Review by brminich — APPROVED** [review]

**Comment by brminich** [comment]
@yosefe, pls review

File: oshmem/shmem/c/shmem_mcs_lock.c
Comment (rakhmets) [inline oshmem/shmem/c/shmem_mcs_lock.c:1]: It is better to replace `sizeof(int)` by `sizeof(variable_name)` in cswap, add, etc.

Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:1]: Actually, swap, add calls are internal MCA calls of shmem_int_*_* calls. sizeof(int) can help with readability than the variable name.

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+/**
+ * @file
+ */
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:13]: removed

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+
+    RUNTIME_CHECK_INIT();
+    /**
+     *  Initializing next pointer to next mask
+     *  Done atomically to avoid races as NEXT pointer
+     *  can be modified by other PEs while acquiring or
+     *  releasing it.
+     */
+    /**
+     * Can make this to be shmem_atomic_set to be safe in non-cc architecutres
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:78]: Fixed

File: oshmem/shmem/c/shmem_set_lock.c [hunk 22]
```diff
@@ -9,7 +11,6 @@
  *
  * $HEADER$
  */
-
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_set_lock.c]: Done

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+    int swap_cond       = 0;
+    int prev_value      = 0;
+    int tval            = 0;
+    int val_tmp         = 0;
+    int nmask           = 0;
+    int a_val           = 0;
+    int retv            = 0;
+
+    /**
+     * Can make atomic fetch to be safe in non-cc architecutres
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:158]: Fixed

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+
+#define MCSQ_TAIL_OWNER(lock_ptr)\
+    (((uintptr_t)(lock_ptr) / sizeof(long)) % shmem_n_pes())
+
+/** Use basic MCS distributed lock algorithm for lock */
+struct lock_t {
+    /** has meaning only on MCSQ_TAIL OWNER */
+    int tail;
+    /** It has meaning on all PEs */
+    /** The next pointer is a combination of the PE ID and  wait signal */
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c]: Fixed

**Review by rakhmets — APPROVED** [review]

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+     * Swap and make me the new tail and update in tail owner
+     * Get the previous tail PE.
+     */
+    retv = MCA_ATOMIC_CALL(swap(oshmem_ctx_default, (void *)tail,
+                        (void*)&prev_tail,
+                        OSHMEM_ATOMIC_PTR_2_INT(&new_tail_req,
+                                                sizeof(new_tail_req)),
+                        sizeof(int), mcs_tail_owner));
+    RUNTIME_CHECK_RC(retv);
+    prev_tailpe = GET_PE(prev_tail);
```
Comment (yosefe) [inline oshmem/shmem/c/shmem_mcs_lock.c:99]: space line before, not after

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+    int next;
+};
+typedef struct lock_t lock_t;
+
+#define NEXT_MASK     0x7FFFFFFFU
+#define SIGNAL_MASK   0x80000000U /** Wait signal mask */
+#define NEXT(A)       (A & NEXT_MASK)
+#define GET_PE(P)     (P & NEXT_MASK) /** Improve readability */
+#define SIGNAL(A)     (A & SIGNAL_MASK)
+#define SET_SIGNAL(A) (A | SIGNAL_MASK)
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:48]: Fixed.

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+    MCA_SPML_CALL(quiet(oshmem_ctx_default));
+
+    /** Signal for setting lock */
+    new_tail_req = SET_SIGNAL(my_pe);
+    /**
+     * Swap and make me the new tail and update in tail owner
+     * Get the previous tail PE.
+     */
+    retv = MCA_ATOMIC_CALL(swap(oshmem_ctx_default, (void *)tail,
+                        (void*)&prev_tail,
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:94]: Fixed.

File: oshmem/shmem/c/shmem_mcs_lock.c (hunk 25 above)
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:99]: Fixed.

File: oshmem/shmem/c/shmem_mcs_lock.c [hunk 28]
```diff
@@ -0,0 +1,234 @@
...
+    retv = MCA_ATOMIC_CALL(cswap(oshmem_ctx_default,
+                (void *)tail, (uint64_t *)&(prev_tail),
+                OSHMEM_ATOMIC_PTR_2_INT(&tmp_cond, sizeof(tmp_cond)),
+                OSHMEM_ATOMIC_PTR_2_INT(&new_tail_req, sizeof(new_tail_req)),
+                sizeof(int), mcs_tail_owner));
+    RUNTIME_CHECK_RC(retv);
+    if (0 == prev_tail) {
+        return 0; /** lock acquired successfully */
+    }
+    return 1; /** lock is set but returning without waiting */
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_mcs_lock.c:233]: Done.

File: oshmem/runtime/oshmem_shmem_params.c [hunk 29]
```diff
@@ -38,6 +39,19 @@ int oshmem_shmem_register_params(void)
                                  MCA_BASE_VAR_SCOPE_READONLY,
                                  &oshmem_shmem_lock_recursive);
 
+    (void) mca_base_var_register("oshmem",
+                                 "oshmem",
+                                 NULL,
+                                 "mcs_lock",
+                                 "Choose mcs locks (default = 0)",
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/runtime/oshmem_shmem_params.c]: MCS is default and removed the default=0 text in help.

File: oshmem/shmem/c/shmem_clear_lock.c [hunk 30]
```diff
@@ -27,5 +30,11 @@
 
 void shmem_clear_lock(volatile long *lock)
 {
-    _shmem_clear_lock((void *)lock, sizeof(long));
+    if (oshmem_shmem_mcs_locks) {
```
[the remark this answers is not in the document]
Comment (vvenkates27) [reply oshmem/shmem/c/shmem_clear_lock.c]: We can probably extend this as we add more locking schemes.

File: oshmem/runtime/oshmem_shmem_params.c [hunk 31]
```diff
@@ -42,15 +42,15 @@ int oshmem_shmem_register_params(void)
     (void) mca_base_var_register("oshmem",
                                  "oshmem",
                                  NULL,
-                                 "mcs_lock",
-                                 "Choose mcs locks (default = 0)",
+                                 "disable_mcs_lock",
```
Comment (yosefe) [inline oshmem/runtime/oshmem_shmem_params.c:45]: better keep it as "enable_mcs_lock". default value = 1. user can set to 0 to disable

Comment (vvenkates27) [reply oshmem/runtime/oshmem_shmem_params.c:45]: Sounds good. Modified now.

**Review by brminich — APPROVED** [review]

**Comment by brminich** [comment]
@vvenkates27 can you please port it to 4.1.x and 5.0 branches?

**Comment by vvenkates27** [comment]
> @vvenkates27 can you please port it to 4.1.x and 5.0 branches?

Submitted pull requests for both branches.
